chalk-solve/src/clauses/builtin_traits/pointee.rs

```diff
@@ -0,0 +1,145 @@
+use crate::rust_ir::{AdtKind, WellKnownTrait};
+use crate::{Interner, RustIrDatabase, TraitRef};
+use chalk_ir::{
+    AliasTy, Floundered, Normalize, ProjectionTy, Substitution, Ty, TyKind, TyVariableKind,
+};
+
+fn push_clauses<I: Interner>(
+    db: &dyn RustIrDatabase<I>,
+    builder: &mut ClauseBuilder<'_, I>,
+    self_ty: Ty<I>,
+    metadata: Ty<I>,
+) {
+    let interner = db.interner();
+    let trait_id = db.well_known_trait_id(WellKnownTrait::Pointee).unwrap();
+    let substitution = Substitution::from1(interner, self_ty);
+    builder.push_fact(TraitRef {
+        trait_id,
+        substitution: substitution.clone(),
+    });
+    let trait_datum = db.trait_datum(trait_id);
+    assert_eq!(
+        trait_datum.associated_ty_ids.len(),
+        1,
+        "Pointee trait should have exactly one associated type, found {:?}",
+        trait_datum.associated_ty_ids
+    );
+    let metadata_id = trait_datum.associated_ty_ids[0];
+    let alias = AliasTy::Projection(ProjectionTy {
+        associated_ty_id: metadata_id,
+        substitution,
+    });
+    builder.push_fact(Normalize {
+        alias,
+        ty: metadata,
+    });
+}
+
+/// Add implicit impl for the `Pointee` trait for all types
+pub fn add_pointee_program_clauses<I: Interner>(
+    db: &dyn RustIrDatabase<I>,
+    builder: &mut ClauseBuilder<'_, I>,
+    self_ty: Ty<I>,
+) -> Result<(), Floundered> {
+    let interner = db.interner();
+    let mut push = |m| push_clauses(db, builder, self_ty.clone(), m);
+    match self_ty.kind(interner) {
+        TyKind::Str | TyKind::Slice(_) => {
+            push(TyKind::Scalar(chalk_ir::Scalar::Uint(chalk_ir::UintTy::Usize)).intern(interner))
+        }
+        TyKind::Array(_, _)
+        | TyKind::Never
+        | TyKind::Closure(_, _)
+        | TyKind::FnDef(_, _)
+        | TyKind::Scalar(_)
+        | TyKind::Raw(_, _)
+        | TyKind::Function(_)
+        | TyKind::InferenceVar(_, TyVariableKind::Float)
+        | TyKind::InferenceVar(_, TyVariableKind::Integer)
+        | TyKind::Generator(_, _)
+        | TyKind::GeneratorWitness(_, _)
+        | TyKind::Ref(_, _, _) => {
+            push(TyKind::Tuple(0, Substitution::empty(interner)).intern(interner))
+        }
+        TyKind::Adt(id, subst) => {
+            let adt_datum = db.adt_datum(*id);
+            if adt_datum.kind != AdtKind::Struct {
+                push(TyKind::Tuple(0, Substitution::empty(interner)).intern(interner));
+                return Ok(());
+            }
+            let last_field_ty = adt_datum
+                .binders
+                .map_ref(|b| b.variants.clone())
+                .substitute(interner, subst)
+                .into_iter()
+                .take(1) // We have a struct so we're guaranteed one variant
+                .flat_map(|mut v| v.fields.pop())
+                .next();
+            push_for_last_field(last_field_ty, db, builder, self_ty);
+        }
+        TyKind::Tuple(_, subst) => {
```

Comment:
I'd probably just break out the unit tuple and unit adt cases separately rather than hiding that logic in `push_for_last_field`